spec/unit/ssl/state_machine_spec.rb

```diff
@@ -61,6 +102,69 @@
       expect(ssl_context[:private_key]).to eq(private_key)
       expect(ssl_context[:client_cert]).to eq(client_cert)
     end
+
+    context 'when exceptions occur' do
+      before :each do
+        allow(cert_provider).to receive(:load_cacerts).and_return(cacerts)
+        allow(cert_provider).to receive(:load_crls).and_return(crls)
+      end
+
+      it 'retries CSR submission' do
+        allow(cert_provider).to receive(:load_private_key).and_return(private_key)
+        allow($stdout).to receive(:puts).with(/Couldn't fetch certificate from CA server; you might still need to sign this agent's certificate/)
+
+        stub_request(:get, %r{puppet-ca/v1/certificate/#{Puppet[:certname]}})
+          .to_return(status: 404).then
+          .to_return(status: 200, body: client_cert.to_pem)
+        stub_request(:put, %r{puppet-ca/v1/certificate_request/#{Puppet[:certname]}})
+          .to_raise(Errno::ECONNREFUSED).then
+          .to_return(status: 200)
+
+        machine.ensure_client_certificate
+
+        expect(@logs).to include(an_object_having_attributes(message: refused_message))
```

Comment:
Not clear how this shows retries...?  I can see from the later test that you can show here you are not getting an error.  Maybe make sure it got the client_cert?

Reply:
I added a commit to verify that the requests are retried